CIP-2551/CIP-2551.md

```diff
@@ -0,0 +1,271 @@
+    * `Ed25519_G_neg :: Ed25519GElement -> Ed25519GElement`
+ * Basic field operations:
+    * `Ed25519_F_add :: Ed25519FElement -> Ed25519FElement -> Ed25519FElement`
+    * `Ed25519_F_mul :: Ed25519FElement -> Ed25519FElement -> Ed25519FElement`
+    * `Ed25519_F_neg :: Ed25519FElement -> Ed25519FElement`
+    * `Ed25519_F_recip :: Ed25519FElement -> Maybe Ed25519FElement`
+ * Vector space operations:
+    * `Ed25519_scale :: Ed25519FElement -> Ed25519GElement -> Ed25519GElement`
+    * `Ed25519_scale_and_add :: [Ed25519FElement] -> [Ed25519GElement] -> Ed25519GElement`
+ * Conversions:
+    * `Ed25519_G_serialise :: Ed25519GElement -> ByteString`
+    * `Ed25519_G_deserialise :: ByteString -> Maybe Ed25519GElement`
+    * `Ed25519_F_serialise :: Ed25519FElement -> ByteString`
+    * `Ed25519_F_deserialise :: ByteString -> Maybe Ed25519FElement`
+    * `Ed25519_F_to_integer :: Ed25519FElement -> Integer`
+    * `Ed25519_F_from_integer :: Integer -> Ed25519FElement`
+ * Equality comparisons:
+    * `eq :: Ed25519GElement -> Ed25519GElement -> bool`
+    * `eq :: Ed25519FElement -> Ed25519FElement -> bool`
+ * Identities:
+    * `Ed25519_G_zero :: Ed25519GElement`
+    * `Ed25519_F_zero :: Ed25519FElement`
+    * `Ed25519_F_one :: Ed25519FElement`
+
+This makes for two new types, 16 new functions, and three new constants.
+
+This CIP writes the Ed25519 elliptic curve group in additive notation.
+The vector space operations consider the Ed25519 curve as a vector
+space over the scalar field `Ed25519FElement`. This implies that there
+is a scalar multiplication operation, which is `Ed25519_scale`. This
+also implies that we can take linear combinations of vectors of curve
+elements, which is what `Ed25519_scale_and_add` does.
+
+The Ed25519 curve is the elliptic curve defined by the equation:
+
+```
+-x^2 + y^2 = 1 - (121665/121666) * x^2 * y^2
+```
+
+This is a twisted Edwards curve.
+
+Field elements are serialized in big endian notation, taking up 32
+bytes each.  This leaves one unused bit, which should be set to zero.
+
+Group elements can be serialized in compressed form (just the
+x-coordinate), or in uncompressed form (both the x-coordinate and
+the y-coordinate).  A group element encoding consists of 65 bytes in
+compressed form, or 129 bytes in uncompressed form. The first byte should
+have all bits except the three least significant bits set to zero. Let
+b0 denote the least significant bit, b1 the second least significant,
+and b2 the third least significant.
+
+ * b0 is 1 if and only if the group element is in compressed form. 
+ * b1 is 1 if and only if the element is the point at infinity.
+   (The rest of the encoding after b1 should be zero in case b1 1 is 1.)
+ * b2 is 1 if and only if the element is in compressed form and not the
+   point at infinity and the point is "positive," meaning that its y
+   coordinate is lexicographically greater than the y coordinate of the
+   other point on the curve with the same x coordinate.
```

Comment:
I now see where the "extra" byte comes in from above, but it still should be 33 in compressed form and 65 in uncompressed form. However, I don't think this is the most efficient representation. We can use a representation that requires only 32 and 64 bytes. Also, due to the fact that we need to cost functions, I would stick to a single serialised representation (possibly the compressed one). Otherwise it would be hard to cost the function as the cost would differ for compressed and uncompressed group elements.